### ISSUE TITLE
NFT contracts with dynamic `get-last-token-id` values are not supported

### ISSUE THREAD

**Issue by 314159265359879** [hirosystems/token-metadata-api#140]
Would you mind having a look at these cases, it looks like not all the NFT's are getting loaded. 
https://github.com/hirosystems/wallet/issues/3290

A user today reported the NFT's from this contract not loading: SP1Q38DQ4EAXEP1YDSAV4C0PBPZF34TTK5NBTAZ0X.a-taste-of-portugal

**Comment by rafaelcr** [comment]
I think I've found the underlying issue... some of these contracts begin with a `get-last-token-id` value of 0 upon deployment, and they're later updated with new mints (or some other method) to reflect the latest token ID. This confuses the Metadata API because it expects the `get-last-token-id` value to return the max number of tokens at the start.

Can you explain a bit more how these contracts are getting generated @314159265359879 ?

**Comment by 314159265359879** [comment]
> Can you explain a bit more how these contracts are getting generated @314159265359879 ?

This particular contract is created with portal for creators at https://create.gamma.io/ 
![image](https://user-images.githubusercontent.com/33360391/226058548-bfd7dcbc-8054-43d6-a82e-414fcc97db40.png)

I see Megapont Apes not displaying either... also because they started with last-id 0? I believe a lot of contracts have started with a lower number and then allowed minting upto a defined maximum. Sometimes the maximum was reached, other times it was not. 
![image](https://user-images.githubusercontent.com/33360391/226058946-be3956c7-0bf6-4b8d-80a0-f0a8e680d66e.png)
https://explorer.stacks.co/txid/SP3D6PV2ACBPEKYJTCMH7HEN02KP87QSP8KTEH335.megapont-ape-club-nft?chain=mainnet

Does this help?

**Comment by svntax** [comment]
The MetaBoy collection also isn't loading in the Hiro wallet, and trying to get token metadata fails with `Token not found`.

For example: https://api.mainnet.hiro.so/metadata/v1/nft/SP2W12MNM4SPV37VZHN4GCDG35G9ACG3FDKK7WF04.MetaBoy/100

I'm guessing it's likely being affected by the same issue, so I'm leaving this comment here.

**Comment by markmhx** [comment]
@rafaelcr I was just talking with @314159265359879 about this and it seems to be a persistent issue for some users. Any updates by chance on its resolution?

**Comment by rafaelcr** [comment]
@markmhx unfortunately this is still blocked by #131 and to fix that one a fairly large refactor is required... so it's still in progress. If there is an urgent token that needs to be shown we could refresh the contract manually, though. Not very sustainable but it'll have to do for now. Let me know.

**Comment by hozzjss** [comment]
Hey @rafaelcr I was wondering if the sip-019 notifications could create new entries in your database, since we're building trajan and the way endorsements work is that they are created on the spot

**Comment by markmhx** [comment]
@314159265359879 curious if you've been seeing more of these incidents lately?

**Comment by rafaelcr** [comment]
Sorry for the late response @hozzjss . Unfortunately, SIP-019 only covers a single token for refresh. There could probably be an amend for that SIP to introduce a new kind of notification that would refresh the entire contract instead of only 1 token, but in the meantime we'll probably need to scan `mint` events somehow to check if the `get-last-token-id` we had in store has been surpassed and do a refresh in that scenario.

**Comment by hozzjss** [comment]
Hey @rafaelcr thanks for the response, I took a look at both the sip 19 text and the code and my own thinking is that maybe it can be like one of the events that are mentioned in sip 19 that way it can incrementally add a token, there's a workaround that im working on atm which is just put a static uint like `u10000` so that the api expects the collection to be that much and get-token-uri would also return a static ipfs uri that is then updated through the sip 19 methods, what do you think would be the best approach?

**Comment by RagnarLifthrasir** [comment]
This issue affects our NFT endorsements on Trajan.App. Our users can only see their NFTs on Trajan. We appreciate you guys following up on this bug.

**Comment by RagnarLifthrasir** [comment]
Hello, @smcclellan would you guys have an update on this issue? Thanks.

**Comment by smcclellan** [comment]
Hey @RagnarLifthrasir, this is scheduled to start work in the coming weeks as it has a few dependencies to tackle first. If there are any tokens you would like manually refreshed in the meantime, please provide them and we can do that as a stop-gap measure.

**Comment by RagnarLifthrasir** [comment]
Thanks @smcclellan we look forward to the update.

**Comment by RagnarLifthrasir** [comment]
Hello @smcclellan just following up. We appreciate you guys working on this. Thanks.

**Comment by bogachevanton** [comment]
Hey @smcclellan @rafaelcr 
Have same problem with contract - SP2ESPYE74G94D2HD9X470426W1R6C2P22B4Z1Q5.skullcoin-souls-nft
Can you please update data manually


First: https://gateway.pinata.cloud/ipfs/QmVNsUvt9osem2R8T8PuXXZu4hsUBRHndHFA38VX4dpWWd/1.json
Last: https://gateway.pinata.cloud/ipfs/QmVNsUvt9osem2R8T8PuXXZu4hsUBRHndHFA38VX4dpWWd/500.json

**Comment by bogachevanton** [comment]
Is there anyone here who can help with this?

**Comment by 314159265359879** [comment]
In the past I have seen @rafaelcr can add it manually.

**Comment by bogachevanton** [comment]
> Done, contract refreshed @bogachevanton 

Thanks you and @314159265359879

**Comment by rafaelcr** [comment]
@314159265359879 yep, done, I've manually refreshed the metadata and it now has all fields filled in: https://api.hiro.so/metadata/v1/ft/SP3K8BC0PPEVCV7NZ6QSRWPQ2JE9E5B6N3PA0KBR9.token-abtc
I believe the API had a problem retrieving the original metadata file when it first saw the token deploy

**Comment by 314159265359879** [comment]
@rafaelcr would you mind manually updating the metadata for SP3D6PV2ACBPEKYJTCMH7HEN02KP87QSP8KTEH335.mega

It looks unavailable when looking at the token in Leather.

**Comment by rafaelcr** [comment]
@314159265359879 sure. I've just refreshed its metadata but the result stayed the same: https://api.hiro.so/metadata/v1/ft/SP3D6PV2ACBPEKYJTCMH7HEN02KP87QSP8KTEH335.mega

Could it be something else causing the problem on the linked issue?

**Comment by 52** [comment]
hey @rafaelcr  👋 - would you mind manually refreshing these:
https://api.hiro.so/metadata/v1/ft/SP2VCQJGH7PHP2DJK7Z0V48AGBHQAW3R3ZW1QF4N.zststx
https://api.hiro.so/metadata/v1/ft/SP2VCQJGH7PHP2DJK7Z0V48AGBHQAW3R3ZW1QF4N.zaeusdc

Doesn't seem to be refresh after calling `token-metadata-update-notify`.

**Comment by rafaelcr** [comment]
sure @52 , done

**Comment by 52** [comment]
@rafaelcr - sorry, seems like there was an issue with the json - can you do one more refresh of both assets?

**Comment by 314159265359879** [comment]
@rafaelcr I see that the token metadata for Velar doesn't include an image when checked via the Hiro API. I think that is because the token-uri was not accessible/available when the contract was deployed. 
https://api.hiro.so/metadata/v1/ft/SP1Y5YSTAHZ88XYK1VPDH24GY0HPX5J4JECTMY4A1.velar-token
![image](https://github.com/hirosystems/token-metadata-api/assets/33360391/0ae5bcf0-efd5-43be-a252-c10184698046)

I see this now:
https://www.velar.co/metadata/velar-token.json
![image](https://github.com/hirosystems/token-metadata-api/assets/33360391/42f5c061-3acf-42d0-a81c-6cefd6dd098a)

![image](https://github.com/hirosystems/token-metadata-api/assets/33360391/67d732db-3766-4c65-864d-705878aea5e3)

Is this the using the right json schema? can it be updated manually now?

**Comment by rafaelcr** [comment]
hi @314159265359879 I've just refreshed the token and I see the image now

**Comment by 314159265359879** [comment]
I am seeing multiple reports of users who have this starnft but it is not displayed, this is a contract with dynamic `get-last-token-id`. 

SPYHQS4RFA4N87RDWNPEWYRYK3C6N2Q8S5CY3XG5.starnft
https://explorer.hiro.so/sandbox/contract-call/SPYHQS4RFA4N87RDWNPEWYRYK3C6N2Q8S5CY3XG5.starnft?chain=mainnet

Token uri's look like this
https://graphigo.prd.galaxy.eco/metadata/SPYHQS4RFA4N87RDWNPEWYRYK3C6N2Q8S5CY3XG5.starnft/1.json

Metadata like this

https://cdn.galxe.com/galaxy/summerspace/ae982bcb-f315-4469-80cf-80a65dcc6f62.png
```
{"name":"test stacks1","description":" Minted by Galxe Campaign: test stacks1","image":"https://cdn.galxe.com/galaxy/summerspace/ae982bcb-f315-4469-80cf-80a65dcc6f62.png","animation_url":"","background_color":"","external_link":"https://galaxy.eco/nft/145986701/SPYHQS4RFA4N87RDWNPEWYRYK3C6N2Q8S5CY3XG5.starnft","owner":"SP162B36N2CPX7DHT26AWF2Z9Z4HHXNQAR2EBRZ0V","attributes":[{"display_type":"date","trait_type":"birthday","value":1710818074}]}
```
Image is accessable
![image](https://github.com/hirosystems/token-metadata-api/assets/33360391/3cae84be-5310-4e53-857a-619b833fca9b)

Could a manual intervention help display these?

**Comment by rafaelcr** [comment]
Sure @314159265359879 I've done a manual refresh of the contract but I'm getting a lot of 404s when retrieving metadata JSONs for a lot of those tokens, example: https://graphigo.prd.galaxy.eco/metadata/SPYHQS4RFA4N87RDWNPEWYRYK3C6N2Q8S5CY3XG5.starnft/87.json

**Comment by rafaelcr** [comment]
Closing this because it will be solved by #146 which is in progress
